extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java

```diff
@@ -236,6 +253,58 @@ public AuthenticatedUser authenticateUser(Credentials credentials)
 
     }
 
+    /**
+     * Returns all custom LDAP attributes on the user currently bound under
+     * the given LDAP connection. The custom attributes are specified in
+     * guacamole.properties.
+     *
+     * @param ldapConnection
+     *     LDAP connection to find the custom LDAP attributes.
+     * @param username
+     *     The username of the user whose attributes are queried.
+     *
+     * @return
+     *     All attributes on the user currently bound under the
+     *     given LDAP connection, as a map of attribute name to
+     *     corresponding attribute value.
+     *
+     * @throws LDAPException
+     *     If an error occurs while searching for the user attributes.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs retrieving the user DN.
+     */
+    private Map<String, String> getLDAPAttributes(LDAPConnection ldapConnection,
+            String username) throws LDAPException, GuacamoleException {
+
+        // Get attributes from configuration information
+        List<String> attrList = confService.getAttributes();
+
+        // If there are no attributes there is no reason to search LDAP
+        if (attrList.size() == 0)
+            return null;
+
+        // Build LDAP query parameters
+        String[] attrArray = attrList.toArray(new String[attrList.size()]);
+        String userDN = getUserBindDN(username);
+
+        // Get LDAP attributes by querying LDAP
+        LDAPEntry userEntry = ldapConnection.read(userDN, attrArray);
+        LDAPAttributeSet attrSet = userEntry.getAttributeSet();
+
+        // Add each attribute into Map
+        Map<String, String> attrMap = new HashMap<String, String>();
+        Iterator attrIterator = attrSet.iterator();
+        while (attrIterator.hasNext()) {
```

Comment:
Might use the Java for-each loop, here - you used it below when dealing with the LDAP attributes in the token - no reason not to use it, here, too! :smile:

Reply:
the reason I did this is because I couldn't get the casting to work in the for-each loop, but I think I got the for-each loop working now.

Reply:
Oh, I see what you mean - yeah, that could be a problem, as those custom classes (particularly ancient ones like the JLDAP ones :smile:) don't always plan nice with the for-each loop.

Reply:
Here is what I am doing now:
       for (Object attrObj : attrMap) {
            LDAPAttribute attr = (LDAPAttribute)attrObj;
            .....
       }

Reply:
Yep, that's fine.  And, depending on what you're doing with it later, you might not even need that assignment, there, you could do:

    String attrName = ((LDAPAttribute) attrObj).getName();
    String attrValue = ((LDAPAttribute) attrObj).getStringValue();

I'm not certain what penalty there is, if any, to doing the cast twice like that, so it may actually make sense to go ahead and assign it to a new `LDAPAttribute` type.  Either way is probably perfectly fine.